### PR TITLE
Feature: Parallel Building

### PR DESCRIPTION
**Pull request by Cregrant**
Here are my performance optimizations that aim to minimize changes to the codebase and do not use any external libraries (although Apache Compress seems like a nice one).
- There is some duplicated code, such as the ApkBuilder and ApkDecoder new fields, which can be extracted to a "global" context created somewhere in the Main class. However, I do not want to make significant changes for just one pull request.
- I did not find any changes in the decompiled code, compression type issues, or any other problems.
- Further performance optimizations can be done by changing the zip compressing library and refactoring the backsmali code (specifically, the synchronized method inside the library).

**Best numbers (30mb apk, 3 dex files, 4 CPU cores) for a 10-minute benchmark:**
_25% means one core load_
| Task              | Configuration     | Time (s) | CPU load|
|-------------------|-------------------|----------|------------|
| decode | original          | 18.05    | 34%        |
| decode | 1thread           | 18.05    | 34%        |
| decode | 2threads          | 15.00    | 40%        |
| decode | 4threads          | 13.40    | 46%        |
| decode | 4th with 2nd commit           | 12.60    | 48%        |
|||||
| encode | original          | 27.50    | 25%        |
| encode | 4threads          | 15.55    | 49%        |
| encode | mergeZip          | 11.45    | 53%        |

### CODE REVIEW

**Comment by iBotPeaches** [comment]
Interesting in timing how we got 1 PR doing ParallelStreams and this one doing an executor service. It what I was basically responding to in this other pr: https://github.com/iBotPeaches/Apktool/pull/3411#issuecomment-1849891110

In short - thanks, let me review this.

File: brut.apktool/apktool-lib/src/main/java/brut/androlib/ApkBuilder.java [hunk 1]
```diff
@@ -78,48 +80,53 @@ public ApkBuilder(Config config, ExtFile apkDir) {
...
-        // this is because Aapt won't add files it doesn't know (ex unknown files)
-        buildUnknownFiles(outFile);
-
-        // we copied the AndroidManifest.xml to AndroidManifest.xml.orig so we can edit it
-        // lets restore the unedited one, to not change the original
-        if (manifest.isFile() && manifest.exists() && manifestOriginal.isFile()) {
-            try {
-                if (new File(mApkDir, "AndroidManifest.xml").delete()) {
-                    FileUtils.moveFile(manifestOriginal, manifest);
+            buildApk(outFile);                              //finish the apk by adding common and unknown files
```
Comment (iBotPeaches) [inline brut.apktool/apktool-lib/src/main/java/brut/androlib/ApkBuilder.java:113]: For all these comments can we either remove them and adapt methods to sustain what the comment would need? and/or just add a leading space so its like:

```
// create classes.dex
```

File: brut.apktool/apktool-lib/src/main/java/brut/androlib/BackgroundWorker.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package brut.androlib;
+
+import java.util.ArrayList;
+import java.util.concurrent.*;
+
+public class BackgroundWorker {
+
+    private static final int THREADS_COUNT = Runtime.getRuntime().availableProcessors();
```
Comment (iBotPeaches) [inline brut.apktool/apktool-lib/src/main/java/brut/androlib/BackgroundWorker.java:8]: Should this be allowed to be changed via like `--jobs`? It seems common in multi-threaded tools to control the amount of threads.

File: brut.apktool/apktool-lib/src/main/java/brut/androlib/ApkBuilder.java [hunk 3]
```diff
@@ -141,13 +148,23 @@ private void buildManifestFile(File manifest, File manifestOriginal) throws Andr
...
+                }
+            } catch (AndrolibException e) {
+                mBuildError.compareAndSet(null, e);
+            }
+        };
+        mWorker.submit(r);
     }
 
-    private void buildNonDefaultSources() throws AndrolibException {
+    private void scheduleBuildNonDefaultSources() throws AndrolibException {
```
Comment (iBotPeaches) [inline brut.apktool/apktool-lib/src/main/java/brut/androlib/ApkBuilder.java:167]: Should we extract the common logic that seems to be the scheduling of the disassembly of dexs into its own method? It seems the single one (classes.dex) and iterating multi-dex is setting up the same run-able to submit.

**Review by iBotPeaches — COMMENTED** [review]
Took a quick static look. I have time this weekend to pull it down and do some testing.

File: brut.apktool/apktool-lib/src/main/java/brut/androlib/BackgroundWorker.java (hunk 2 above)
Comment (Cregrant) [reply brut.apktool/apktool-lib/src/main/java/brut/androlib/BackgroundWorker.java:8]: No problem. Should we then delete `BackgroundWorker(int threads)` method?

Comment (Cregrant) [reply brut.apktool/apktool-lib/src/main/java/brut/androlib/BackgroundWorker.java:8]: > Should this be allowed to be changed via like `--jobs`? It seems common in multi-threaded tools to control the amount of threads.

No problem. Should we then delete `BackgroundWorker(int threads)` method?

**Comment by Cregrant** [comment]
Another suggestion is to change the Gradle distribution from `all` to `bin` if you didn't choose `all` intentionally.

**Comment by iBotPeaches** [comment]
> Another suggestion is to change the Gradle distribution from `all` to `bin` if you didn't choose `all` intentionally.

We did swap from bin -> all a few years ago for some reason. Forgot why.

**Review by iBotPeaches — APPROVED** [review]
I ran some tests and I like it. I'll put in another PR shortly to:

 * Make jobs configurable
 * Explain how many jobs are running in info message at start of build
 * Remove unused method in ZipUtils

**Comment by Cregrant** [comment]
Thanks! I thought about defining the thread count and found that setting the number in the Main class (or setting it to 1 if a debug flag is set) via a public field is the best solution. This is because you probably won't have any reason to change it in the middle of decompiling.

**Pull request merged**